File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null) {return null;}
        if (l1 == null) {return l2;}
        if (l2 == null) {return l1;}

        int countl1 = 0;
        int countl2 = 0;

        ListNode temp = l1;

        while (temp != null) {
            countl1++;
            temp = temp.next;
        }

        temp = l2;

        while (temp != null) {
            countl2++;
            temp = temp.next;
        }

        ListNode l3 = new ListNode();

        if (countl1 >= countl2) {
            l3 = l1;
            temp = l3;
            while (l2 != null) {
                temp.val += l2.val;
                temp = temp.next;
                l2 = l2.next;
            }
        }
        else {
            l3 = l2;
            temp = l3;
            while (l1 != null) {
                temp.val += l1.val;
                temp = temp.next;
                l1 = l1.next;
            }
        }
        temp = l3;
        while (temp != null) {
            if (temp.val > 9) {
                if (temp.next != null) {
                    temp.val %= 10;
                    temp.next.val += 1;
                }
                else {
                    temp.val %= 10;
                    ListNode aux = new ListNode(1);
                    temp.next = aux;
                }
            }
            temp = temp.next;
        }
        return l3;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* resp = new ListNode(l1->val + l2->val);
        ListNode* check = resp;
        l1 = l1->next;
        l2 = l2->next;
        while (l1 || l2) {
            if (l2 == NULL && l1 != NULL) {check->next = l1; break;}
            if (l1 == NULL && l2 != NULL) {check->next = l2; break;}
            ListNode* temp = new ListNode(l1->val + l2->val);
            check->next = temp;
            check = check->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        check = resp;
        while (check) {
            if (check->val > 9 && check->next == NULL) {
                check->val %= 10;
                ListNode* temp = new ListNode(1);
                check->next = temp;
            }
            else if (check->val > 9) {
                check->val %= 10;
                check->next->val++;
            }
            check = check->next;
        }
        return resp;
    }
};